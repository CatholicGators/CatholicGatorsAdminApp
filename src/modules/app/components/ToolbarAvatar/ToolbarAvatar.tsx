import React from "react"

import { createStyles, withStyles, Theme } from "@material-ui/core"
import IconButton from "@material-ui/core/IconButton"
import Avatar from "@material-ui/core/Avatar"
import MenuItem from "@material-ui/core/MenuItem"
import Menu from "@material-ui/core/Menu"
import Button from "@material-ui/core/Button"
import CircularProgress from "@material-ui/core/CircularProgress"

export const styles = (theme: Theme) =>
    createStyles({
        progress: {
            margin: 12,
            color: theme.palette.secondary.main,
        },
    })

type Props = {
    classes: any
    user: any
    login: () => any
    logout: () => any
}

type State = {
    anchorEl: string | null
}

export class ToolbarAvatar extends React.Component<Props, State> {
    state = {
        anchorEl: null,
    };

    handleMenu(event) {
        this.setState({
            anchorEl: event.currentTarget,
        })
    }

    handleClose() {
        this.setState({
            anchorEl: null,
        })
    }

    handleLogin() {
        this.props.login()
    }

    handleLogout() {
        this.props.logout()
        this.handleClose()
    }

    render() {
        const { anchorEl } = this.state
        const { classes, user } = this.props

        switch (this.props.user) {
            case undefined:
                return (
                    <CircularProgress
                        id="spinner"
                        className={classes ? classes.progress : null}
                    />
                )

            case null:
                return (
                    <Button
                        id="login-btn"
                        color="inherit"
                        variant="outlined"
                        onClick={this.handleLogin.bind(this)}
                    >
                        Login
                    </Button>
                )

            default:
                return (
                    <div>
                        <IconButton
                            id="avatar-btn"
                            aria-owns={anchorEl ? "menu" : undefined}
                            aria-haspopup="true"
                            onClick={this.handleMenu.bind(this)}
                            color="inherit"
                        >
                            <Avatar src={user.photoUrl} />
                        </IconButton>
                        <Menu
                            id="menu"
                            anchorEl={anchorEl}
                            anchorOrigin={{
                                vertical: "top",
                                horizontal: "right",
                            }}
                            transformOrigin={{
                                vertical: "top",
                                horizontal: "right",
                            }}
                            open={!!anchorEl}
                            onClose={this.handleClose.bind(this)}
                        >
                            <MenuItem
                                id="logout"
                                onClick={this.handleLogout.bind(this)}
                            >
                                Logout
                            </MenuItem>
                        </Menu>
                    </div>
                )
        }
    }
}

export default withStyles(styles)(ToolbarAvatar)
