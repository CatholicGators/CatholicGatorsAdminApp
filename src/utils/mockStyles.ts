import { createMuiTheme } from '@material-ui/core'

export default (styles) => Object.keys(styles(createMuiTheme())).reduce((acc, curr) => ({...acc, [curr]: "test"}), {})