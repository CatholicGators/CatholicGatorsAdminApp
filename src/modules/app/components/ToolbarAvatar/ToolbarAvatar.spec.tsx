import React from "react"
import { shallow } from "enzyme"
import toJson from "enzyme-to-json"

import { ToolbarAvatar, styles } from "./ToolbarAvatar"
import mockStyles from "utils/mockStyles"

describe("ToolbarAvatar", () => {
    let props, wrapper

    beforeEach(() => {
        props = {
            classes: mockStyles(styles),
            user: {},
            login: jest.fn(),
            logout: jest.fn(),
        }
        wrapper = shallow(<ToolbarAvatar {...props} />)
    })

    it("should match snapshot", () => {
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    describe("user is loading", () => {
        beforeEach(() => {
            wrapper.setProps({ user: undefined })
        })

        it("should only show a spinner", () => {
            expect(wrapper.exists("#spinner")).toBe(true)
            expect(wrapper.exists("#login-btn")).toBe(false)
            expect(wrapper.exists("#avatar-btn")).toBe(false)
        })
    })

    describe("user is not logged in", () => {
        beforeEach(() => {
            wrapper.setProps({ user: null })
        })

        it("should only show the login button", () => {
            expect(wrapper.exists("#spinner")).toBe(false)
            expect(wrapper.exists("#login-btn")).toBe(true)
            expect(wrapper.exists("#avatar-btn")).toBe(false)
        })

        it("calls googleSignIn()", () => {
            wrapper.find("#login-btn").simulate("click")

            expect(props.login).toHaveBeenCalled()
        })
    })

    describe("user is logged in", () => {
        const testTarget = "test"

        beforeEach(() => {
            wrapper
                .find("#avatar-btn")
                .simulate("click", { currentTarget: testTarget })
        })

        it("should only show the avatar", () => {
            expect(wrapper.exists("#spinner")).toBe(false)
            expect(wrapper.exists("#login-btn")).toBe(false)
            expect(wrapper.exists("#avatar-btn")).toBe(true)
        })

        it("clicking avatar sets anchorEl", () => {
            expect(wrapper.state("anchorEl")).toBe(testTarget)
        })

        describe("logout", () => {
            it("closes the menu", () => {
                wrapper.find("#logout").simulate("click")

                expect(wrapper.state("anchorEl")).toBe(null)
            })

            it("calls signOut()", () => {
                wrapper.find("#logout").simulate("click")

                expect(props.logout).toHaveBeenCalled()
            })
        })
    })
})
