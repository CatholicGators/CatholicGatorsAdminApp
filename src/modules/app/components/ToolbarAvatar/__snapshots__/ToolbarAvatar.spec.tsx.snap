// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolbarAvatar should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ToolbarAvatar
    login={[MockFunction]}
    logout={[MockFunction]}
    user={Object {}}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <WithStyles(IconButton)
          aria-haspopup="true"
          color="inherit"
          id="avatar-btn"
          onClick={[Function]}
        >
          <WithStyles(Avatar) />
        </WithStyles(IconButton)>,
        <WithStyles(Menu)
          anchorEl={null}
          anchorOrigin={
            Object {
              "horizontal": "right",
              "vertical": "top",
            }
          }
          id="menu"
          onClose={[Function]}
          open={false}
          transformOrigin={
            Object {
              "horizontal": "right",
              "vertical": "top",
            }
          }
        >
          <WithStyles(MenuItem)
            id="logout"
            onClick={[Function]}
          >
            Logout
          </WithStyles(MenuItem)>
        </WithStyles(Menu)>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "aria-haspopup": "true",
          "aria-owns": undefined,
          "children": <WithStyles(Avatar) />,
          "color": "inherit",
          "id": "avatar-btn",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "src": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "anchorEl": null,
          "anchorOrigin": Object {
            "horizontal": "right",
            "vertical": "top",
          },
          "children": <WithStyles(MenuItem)
            id="logout"
            onClick={[Function]}
          >
            Logout
          </WithStyles(MenuItem)>,
          "id": "menu",
          "onClose": [Function],
          "open": false,
          "transformOrigin": Object {
            "horizontal": "right",
            "vertical": "top",
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Logout",
            "id": "logout",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Logout",
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <WithStyles(IconButton)
            aria-haspopup="true"
            color="inherit"
            id="avatar-btn"
            onClick={[Function]}
          >
            <WithStyles(Avatar) />
          </WithStyles(IconButton)>,
          <WithStyles(Menu)
            anchorEl={null}
            anchorOrigin={
              Object {
                "horizontal": "right",
                "vertical": "top",
              }
            }
            id="menu"
            onClose={[Function]}
            open={false}
            transformOrigin={
              Object {
                "horizontal": "right",
                "vertical": "top",
              }
            }
          >
            <WithStyles(MenuItem)
              id="logout"
              onClick={[Function]}
            >
              Logout
            </WithStyles(MenuItem)>
          </WithStyles(Menu)>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "aria-haspopup": "true",
            "aria-owns": undefined,
            "children": <WithStyles(Avatar) />,
            "color": "inherit",
            "id": "avatar-btn",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "src": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "anchorEl": null,
            "anchorOrigin": Object {
              "horizontal": "right",
              "vertical": "top",
            },
            "children": <WithStyles(MenuItem)
              id="logout"
              onClick={[Function]}
            >
              Logout
            </WithStyles(MenuItem)>,
            "id": "menu",
            "onClose": [Function],
            "open": false,
            "transformOrigin": Object {
              "horizontal": "right",
              "vertical": "top",
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Logout",
              "id": "logout",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Logout",
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
